code/go/0chain.net/smartcontract/zcnsc/stakepool.go

```diff
@@ -0,0 +1,313 @@
+	}
+
+	transfer := state.NewTransfer(sscID, clientID, dp.Balance)
+	if err := balances.AddTransfer(transfer); err != nil {
+		return false, err
+	}
+
+	sp.Pools[poolID].Balance = 0
+	sp.Pools[poolID].Status = spenum.Deleting
+
+	return true, nil
+}
+
+//
+// smart contract methods
+//
+
+// getStakePool of given authorizer
+func (zcn *ZCNSmartContract) getStakePool(authorizerID datastore.Key, balances cstate.StateContextI) (sp *StakePool, err error) {
+	sp = newStakePool()
+	err = balances.GetTrieNode(stakePoolKey(zcn.ID, authorizerID), sp)
+	if err != nil {
+		return nil, err
+	}
+
+	return sp, nil
+}
+
+// initial or successive method should be used by add_authorizer
+
+// SC functions
+
+// get existing stake pool or create new one not saving it
+func (zcn *ZCNSmartContract) getOrUpdateStakePool(
+	gn *GlobalNode,
+	authorizerId datastore.Key,
+	providerType spenum.Provider,
+	settings stakepool.StakePoolSettings,
+	balances cstate.StateContextI,
+) (*StakePool, error) {
+	if err := validateStakePoolSettings(settings, gn); err != nil {
+		return nil, fmt.Errorf("invalid stake_pool settings: %v", err)
+	}
+
+	changed := false
+
+	// the stake pool can be created by related validator
+	sp, err := zcn.getStakePool(authorizerId, balances)
+	if err != nil {
+		if err != util.ErrValueNotPresent {
+			return nil, fmt.Errorf("unexpected error: %v", err)
+		}
+		sp = newStakePool()
+		sp.Minter = cstate.MinterStorage
+		sp.Settings.DelegateWallet = settings.DelegateWallet
+		changed = true
+	}
+
+	if sp.Settings.MinStake != settings.MinStake {
+		sp.Settings.MinStake = settings.MinStake
+		changed = true
+	}
+
+	if sp.Settings.MaxStake != settings.MaxStake {
+		sp.Settings.MaxStake = settings.MaxStake
+		changed = true
+	}
+
+	if sp.Settings.ServiceCharge != settings.ServiceCharge {
+		sp.Settings.ServiceCharge = settings.ServiceCharge
+		changed = true
+	}
+
+	if sp.Settings.MaxNumDelegates != settings.MaxNumDelegates {
+		sp.Settings.MaxNumDelegates = settings.MaxNumDelegates
+		changed = true
+	}
+
+	if changed {
+		return sp, nil
+	}
+
+	return nil, fmt.Errorf("no changes have been made to stakepool for authorizerID (%s)", authorizerId)
+}
+
+func validateStakePoolSettings(poolSettings stakepool.StakePoolSettings, conf *GlobalNode) error {
+	err := conf.validateStakeRange(poolSettings.MinStake, poolSettings.MaxStake)
+	if err != nil {
+		return err
+	}
+	if poolSettings.ServiceCharge < 0.0 {
+		return errors.New("negative service charge")
+	}
+	if poolSettings.MaxNumDelegates <= 0 {
+		return errors.New("num_delegates <= 0")
+	}
+
+	return nil
+}
+
+func (gn *GlobalNode) validateStakeRange(min, max state.Balance) (err error) {
+	if min < gn.MinStakeAmount {
+		return fmt.Errorf("min_stake is less than allowed by SC: %v < %v", min, gn.MinStakeAmount)
+	}
+	if max < min {
+		return fmt.Errorf("max_stake less than min_stake: %v < %v", min, max)
+	}
+
+	return
+}
+
+func (zcn *ZCNSmartContract) DistributeRewards(
+	t *transaction.Transaction,
```

Comment:
Can't you use the [DistributeRewards](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/stakepool/stakepool.go#L170) from the stakepool.StakePool object. So this can be removed.

Reply:
OK